### PR TITLE
Structural plasticity

### PR DESCRIPTION
**Pull request by pabogdan**
PyNN 9.5.2+ support for structural plasticity

Depends on https://github.com/SpiNNakerManchester/sPyNNaker/pull/460

### CODE REVIEW

File: spynnaker8/models/synapse_dynamics/synapse_dynamics_structural.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    import SynapseDynamicsStructural as _BaseClass
+from spinn_front_end_common.utilities import globals_variables
+import numpy as np
+
+from spynnaker8.models.synapse_dynamics import SynapseDynamicsStatic
+
+
+class SynapseDynamicsStructural(_BaseClass):
+    __slots__ = [
+        "_f_rew",
```
Comment (alan-stokes) [inline spynnaker8/models/synapse_dynamics/synapse_dynamics_structural.py]: comment the variables stating what they do/are

File: spynnaker8/models/synapse_dynamics/synapse_dynamics_structural.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        "_p_form_forward",
+        "_p_form_lateral",
+        "_p_elim_dep",
+        "_p_elim_pot",
+        "_grid",
+        "_random_partner",
+        "_connections",
+        "_actual_row_max_length"]
+
+    def __init__(self, stdp_model=None, f_rew=10 ** 4, weight=0, delay=1,
```
Comment (alan-stokes) [inline spynnaker8/models/synapse_dynamics/synapse_dynamics_structural.py]: all these variables are magic to me, please create named vairables that explain what they do/are

File: spynnaker8/models/synapse_dynamics/synapse_dynamics_structural.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+                 s_max=32, sigma_form_forward=2.5, sigma_form_lateral=1,
+                 p_form_forward=0.16, p_form_lateral=1,
+                 p_elim_dep=0.0245, p_elim_pot=1.36 * 10 ** -4,
+                 grid=np.array([16, 16]), lateral_inhibition=0,
+                 random_partner=False,
+                 seed=None):
+        # if stdp_model and isinstance(stdp_model, SynapseDynamicsSTDP):
```
Comment (alan-stokes) [inline spynnaker8/models/synapse_dynamics/synapse_dynamics_structural.py]: new line before this

File: spynnaker8/models/synapse_dynamics/synapse_dynamics_structural.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
+from spynnaker.pyNN.models.neuron.synapse_dynamics \
+    import SynapseDynamicsStructural as _BaseClass
```
Comment (alan-stokes) [inline spynnaker8/models/synapse_dynamics/synapse_dynamics_structural.py]: why the rename? it obscures the readability, remove the rename and keep as SynapseDynamicsStructural

File: spynnaker8/models/synapse_dynamics/synapse_dynamics_structural.py (hunk 1 above)
Comment (alan-stokes) [reply spynnaker8/models/synapse_dynamics/synapse_dynamics_structural.py]: most of these slots are not slots for this file, but for the SynapseDynamicsStructural. I dont think yuo need to list these, as they should be slotted in SynapseDynamicsStructural. you only ened to slot self. params of this class

Comment (alan-stokes) [reply spynnaker8/models/synapse_dynamics/synapse_dynamics_structural.py]: as in weights and delays

File: spynnaker8/models/synapse_dynamics/synapse_dynamics_structural.py (hunk 2 above)
Comment (pabogdan) [reply spynnaker8/models/synapse_dynamics/synapse_dynamics_structural.py]: I don't know what this means. Where do I create these names variables?

File: spynnaker8/models/synapse_dynamics/synapse_dynamics_structural.py (hunk 4 above)
Comment (pabogdan) [reply spynnaker8/models/synapse_dynamics/synapse_dynamics_structural.py]: See https://github.com/SpiNNakerManchester/sPyNNaker8/blob/11f2fe1182f744c9d4cfff0299c95e3e97c53359/spynnaker8/models/synapse_dynamics/synapse_dynamics_static.py#L1-L2

Comment (alan-stokes) [reply spynnaker8/models/synapse_dynamics/synapse_dynamics_structural.py]: thats my point. i had to go up there. its not easy readability, and is not any context, youve actively made the code less readable with that change

File: spynnaker8/models/synapse_dynamics/synapse_dynamics_structural.py (hunk 2 above)
Comment (alan-stokes) [reply spynnaker8/models/synapse_dynamics/synapse_dynamics_structural.py]: this is the slots, which i dont think you need to actually fill in. Plus the numebrs in the __init__. what does 10**4 gotta do with f_rew, what is it? 

on the same point where we dont do timeout=0.5 we do timeout=TIMEOUT_IN_SECONDS_FOR_BLAH

**Comment by rowleya** [comment]
Replaced with #244.  To be deleted once this has been merged...

**Pull request merged**